server/transport.go

```diff
@@ -0,0 +1,169 @@
+	clients   map[multiraft.NodeID]*rpc.Client
+}
+
+// NewRPCTransport creates a new RPCTransport with existing gossip and rpc server.
+func NewRPCTransport(gossip *gossip.Gossip, rpcServer *crpc.Server) (multiraft.Transport, error) {
+	t := &RPCTransport{
+		gossip:    gossip,
+		rpcServer: rpcServer,
+		servers:   make(map[multiraft.NodeID]multiraft.ServerInterface),
+		clients:   make(map[multiraft.NodeID]*rpc.Client),
+	}
+
+	err := t.rpcServer.RegisterName("MultiRaft", (*transportRPCServer)(t))
+	if err != nil {
+		return nil, err
+	}
+
+	return t, nil
+}
+
+// transportServer is a type alias to separate RPC methods
+// (which net/rpc finds via reflection) from others.
+type transportRPCServer RPCTransport
+
+// RaftMessage proxies the incoming request to the listening server interface.
+func (t *transportRPCServer) RaftMessage(protoReq *proto.RaftMessageRequest,
+	resp *proto.RaftMessageResponse) error {
+	// Convert from proto to internal formats.
+	req := &multiraft.RaftMessageRequest{GroupID: protoReq.GroupID}
+	if err := req.Message.Unmarshal(protoReq.Msg); err != nil {
+		return err
+	}
+
+	t.mu.Lock()
+	server, ok := t.servers[multiraft.NodeID(req.Message.To)]
+	t.mu.Unlock()
+
+	if ok {
+		return server.RaftMessage(req, &multiraft.RaftMessageResponse{})
+	}
+
+	return util.Errorf("Unable to proxy message to node: %d", req.Message.To)
+}
+
+// Listen registers a ServerInterface to be proxied messages.
+func (t *RPCTransport) Listen(id multiraft.NodeID, server multiraft.ServerInterface) error {
+	t.mu.Lock()
+	t.servers[id] = server
+	t.mu.Unlock()
+	return nil
+}
+
+// Stop unregisters the server id from getting messages.
+func (t *RPCTransport) Stop(id multiraft.NodeID) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	delete(t.servers, id)
+}
+
+func (t *RPCTransport) getClient(id multiraft.NodeID) (*rpc.Client, error) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	client, ok := t.clients[id]
+	if ok {
+		return client, nil
+	}
+
+	nodeIDKey := gossip.MakeNodeIDGossipKey(proto.NodeID(id))
+	info, err := t.gossip.GetInfo(nodeIDKey)
+	if info == nil || err != nil {
+		return nil, util.Errorf("Unable to lookup address for node: %d. Error: %s", id, err)
+	}
+	address := info.(net.Addr)
+
+	conn, err := net.Dial("tcp", address.String())
+	if err != nil {
+		return nil, err
+	}
+	client = rpc.NewClientWithCodec(codec.NewClientCodec(conn))
+
+	t.clients[id] = client
+
+	return client, err
+}
+
+// Send a message to the specified Node id.
+func (t *RPCTransport) Send(id multiraft.NodeID, req *multiraft.RaftMessageRequest) error {
+	// Convert internal to proto formats.
+	protoReq := &proto.RaftMessageRequest{GroupID: req.GroupID}
+	var err error
+	if protoReq.Msg, err = req.Message.Marshal(); err != nil {
+		return err
+	}
+
+	client, err := t.getClient(id)
+	if err != nil {
+		return err
+	}
+	call := client.Go(raftMessageName, protoReq, &proto.RaftMessageResponse{}, nil)
+	select {
+	case <-call.Done:
```

Comment:
Why special case this? The call failing synchronously isn't a common enough occurrence that this warrants two execution pathways. Just always fire-and-forget.